backend/src/org/commcare/session/CommCareSession.java

```diff
@@ -344,14 +344,52 @@ public Menu getMenu(String id) {
         return null;
     }
 
+    /**
+     * When StackFrameSteps are parsed, those that are "datum" operations will be marked as type
+     * "unknown". When we encounter a StackFrameStep of unknown type at runtime, we need to
+     * determine whether it should be interpreted as STATE_DATUM_COMPUTED or STATE_COMMAND_ID.
+     * This primarily affects the behavior of stepBack().
+     *
+     * The logic being employed is: If there is a previous step on the stack whose entries would
+     * have added this command, interpret it as a command. Otherwise, interpret it as a computed
+     * datum.
+     */
+    private String guessUnknownType(StackFrameStep popped){
+        String poppedId = popped.getId();
+        for (StackFrameStep stackFrameStep: frame.getSteps()){
+            String commandId = stackFrameStep.getId();
+            Vector<Entry> entries = getEntriesForCommand(commandId);
+            for (Entry entry: entries){
+                String childCommand = entry.getCommandId();
+                if (childCommand.equals(poppedId)) {
+                    return SessionFrame.STATE_COMMAND_ID;
+                }
+            }
+        }
+        return SessionFrame.STATE_DATUM_COMPUTED;
+    }
+
+    private boolean popNext(EvaluationContext evalContext){
```

Comment:
Can you rename this to better describe what it does? It's not actually popping anything off, just checking the thing that was popped. Something like `lastPoppedWasComputed()`?